### DISCUSSION TITLE
Nodal variables do not have gradients at nodes

### DISCUSSION THREAD

**Discussion by jin0123456** [idaholab/moose#25894] [Q&A General]
I made a custom auxiliary variable and used `MatchedValueBC` to give the boundary conditions of a variable. The error is 'Nodal variables do not have gradients at nodes' which I suppose happened in the auxkernel.
Here is .C of the auxkernel,
```
#include "CladOutsideTemp.h"

registerMooseObject("WorkerApp", CladOutsideTemp);

InputParameters
CladOutsideTemp::validParams()
{
  InputParameters params = AuxKernel::validParams();

  params.addCoupledVar("temperature", "Coupled temperature");
  params.addParam<PostprocessorName>("P", "Internal pressure");
	params.addClassDescription("Compute clad outer temperature");

  return params;
}

CladOutsideTemp::CladOutsideTemp(const InputParameters & parameters)
  : AuxKernel(parameters),
  _temperature(coupledValue("temperature")),
  _grad_t(coupledGradient("temperature")),
  _T_sat(getMaterialProperty<Real>("T_sat")),
  _internal_pressure(getPostprocessorValue("P"))
{
}

Real
CladOutsideTemp::computeValue()
{
	return _T_sat[_qp] + 7.91 * std::exp(-_internal_pressure/1e5/62.) * std::pow((-(6.308 + 2.716e-2 * _temperature[_qp] - 7.301e-6 * std::pow(_temperature[_qp], 2)) * _grad_t[_qp](0)/1000000.), 0.25);
}
```
The input,
```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
  coord_type = RZ
[]

[Problem]
  solve = true
[]
[Variables]
  [convected]
    order = FIRST
    family = MONOMIAL
  []
[]
[AuxVariables]
  [integral_value]
    order = CONSTANT
    family = MONOMIAL
  []
  [h_rise]
    order = CONSTANT
    family = MONOMIAL
  []
  [h]
    order = CONSTANT
    family = MONOMIAL
  []
  [temp]
    order = FIRST
    family = LAGRANGE
  []
  [outer_temp]
    order = FIRST
    family = LAGRANGE
  []
  [T]
    order = CONSTANT
    family = MONOMIAL
  []
  [p]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [integral]
    type = CoolantPerAux
    temp = temp
    variable = integral_value
  []
  [h_rise]
    type = SpatialUserObjectAux
    variable = h_rise
    execute_on = 'initial linear'
    user_object = layered_integral
  []
  [h]
    type = ParsedAux
    variable = h
    #coupled_variables = 'h_rise'
    expression = '1297.86e3 + 4000 * y'
    use_xyzt = true
  []
  [temperature]
    type = ParsedAux
    variable = temp
    expression = '600. - 200. * x'
    use_xyzt = true
  []
  [T]
    type = MaterialRealAux
    property = T
    variable = T
    boundary = right
  []
  [pressure]
    type = ParsedAux
    variable = p
    expression = '15.5e6'
  []
  [outer_temp]
    type = CladOutsideTemp
    temperature = temp
    P = internal_pressure
    variable = outer_temp
  []
[]
[Kernels]
  [diff]
    type = Diffusion
    variable = convected
  []
[]

[FluidProperties]
  [water]
    type = Water97FluidProperties
  []
[]

[Materials]
  [fp_mat]
    type = FluidPropertiesMaterialPHNEW
    pressure = p
    h = h
    fp = water
  []
[]

[UserObjects]
  [layered_integral]
    type = LayeredSideIntegral
    direction = y
    num_layers = 50
    variable = integral_value
    execute_on = 'initial linear'
    cumulative = true
    boundary = right
  []
[]

[BCs]
  [left]
    type = DirichletBC
    variable = convected
    boundary = left
    value = 0
  []
  [right]
    type = MatchedValueBC
    variable = convected
    boundary = right
    v = outer_temp
  []
[]

[Postprocessors]
  [T_sat]
    type = ElementAverageMaterialProperty
    mat_prop = T_sat
  []
  [internal_pressure]
    type = ConstantPostprocessor
    value = 1e6
  []
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'

[]

[Outputs]
  exodus = true
  checkpoint = true
[]
```
So which step leads to the error?

**Comment by GiudGiud** [comment]
Hello

This 
```
  _grad_t(coupledGradient("temperature")),
```
cant be done like this if temperature is a nodal variable. 
since this is an auxkernel, are you using this for output?

Guillaume

**Reply by jin0123456** [reply]
i create this aux to give boundary of another variable like the input file. And I want to call
the gradient of temperature in x direction, how to achieve this in auxkernel?

**Reply by GiudGiud** [reply]
You could project to an elemental variable. (use an element variable for the auxiliary variable being written)

there will be a projection error but it wont be as bad as figuring out which value to use for discontinuous derivatives?

**Reply by jin0123456** [reply]
If just for output, it's ok for elemental variable. But `MatchedValueBC` needs a nodal variable to supply, there is another error if i make it a elemental variable.

**Reply by lindsayad** [reply]
You could project the elemental variable, which represents the gradient projection, onto a nodal variable and then use that in the `MatchedValueBC`, so two projections

**Reply by GiudGiud** [reply]
Two projections can work
@hugary1995 is there some nodal patch equivalent code to choose a derivative for the node? Or to average the derivatives. i know it might not be a great idea

**Reply by jin0123456** [reply]
So is the gradient supposed to be an elemental variable no matter what order is the temperature? And I wonder how to project the elemental variable onto a nodal variable?
Thanks.

**Reply by dschwen** [reply]
@GiudGiud, this is not a bad idea at all. @hugary1995 implemented a nodal patch recovery object. I'm currently working on a proof of concept to use this code to perform nodal projection of stateful material properties. This will allow is to interpolate stateful properties at qudrature point locations that change between timestep (as is required in mortar for example).

**Reply by richmondodufisan** [reply]
Just piggybacking on this as I have the same issue. I don't know if it is worth creating a new thread for.

Just like OP, I initially used MatchedValueBC. Since my BC depends on nodal variables and their gradients, I decided to first calculate the complicated expression within a custom Material file and output it as a Material, which is by nature, defined on the quadrature points i.e is elemental. Then I used the Nodal Patch Recovery to extrapolate to the nodes. A double projection basically, as suggested earlier by @lindsayad (if I understood correctly). The simulation ran, but did not converge.

I then tried something else- created a custom Dirichlet BC (https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex04_bcs.html) where I would evaluate the complicated expression, which seemed to make more sense to do. But then, I could not couple the needed gradients- I got the same error as the title of this thread. I tried calculating the gradients with a Material file, and declaring them as MaterialProperties, but the NodalBC object could not consume them. And so I projected them with Patch Recovery as nodal aux variables and then coupled them to the BC. Again, the simulation ran, but did not converge.

The errors of both attempts were on the same order, I don't know if that means anything. I will still have to properly go through the rest of the code to check for implementation errors, but I wanted to share in case you had any thoughts.

Also, in this scenario, _**would you set execute_on to TIMESTEP_END, TIMESTEP_BEGIN, or NONLINEAR**_? I thought it would be NONLINEAR so that it would solve concurrently with the system of equations. But using NONLINEAR made the simulation abort almost instantly, compared to with TIMESTEP_END, where it still attempted to solve.

**Reply by hugary1995** [reply]
Couple of options here:
1. At each node, grab all the neighboring nodes and estimate the gradient.
2. At each node, grab all the neighboring quadrature points and estimate the gradient. This is the nodal patch recovery technique.
3. Instead of MatchedValueBC, use an integrated BC. An integrated BC operates on quadrature points, and so will have access to both gradients and (interpolated) nodal values.

My suggestion for your particular use case is 3.

Note 3 will give you approximation error as is usual with Galerkin. The error diminishes upon mesh refinement.

**Reply by richmondodufisan** [reply]
I was considering using an IntegratedBC but wasn't quite clear on the documentation https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex04_bcs.html

For the NodalBC, within `computeQpResidual()` the residual is literally u - whatever_u_equals.

As for the IntegratedBC, it works a bit differently doesn't it? The residual is working with the gradient of u, and not u. Right? Or how would I go about implement u=expression with it? In my system I am not dealing with the fluxes.

Also, how would you grab the neighboring nodes in each direction if I wanted to try option 1? What is the syntax for that?

**Reply by lindsayad** [reply]
IntegratedBCs are tied to the surface terms in your weak form

**Reply by richmondodufisan** [reply]
I see and yeah, the PDE in question does not have a surface term so I guess that doesn't work for my case. Also, I got one timestep to converge yesterday, but the simulation immediately failed after. So I guess the issue isn't with execute_on NONLINEAR vs TIMESTEP_END as I said before. It's simply failing whenever it tries to execute. I found this:

> AuxVariables currently do not contain the Automatic Differentiation information about the derivatives. Using this object in the solve, to compute a material property for example, will lead to an imperfect Jacobian which can impact convergence.

Which I guess could explain so, as everything else in my simulation uses AD. Is there a way around this?

**Reply by richmondodufisan** [reply]
I am inheriting from NodalBC instead of ADNodalBC for my custom BC because the latter, for some reason, could not recognize _u[_qp] (said no match for operator "["  or something like that).

**Reply by lindsayad** [reply]
> I see and yeah, the PDE in question does not have a surface term so I guess that doesn't work for my case.

This is pretty rare. Does the strong form of your PDE not have any divergence operators?

> I am inheriting from NodalBC instead of ADNodalBC for my custom BC because the latter, for some reason, could not recognize _u[_qp] (said no match for operator "[" or something like that).

In `ADNodalBC`, you just use `_u` instead of `_u[_qp]`

**Reply by richmondodufisan** [reply]
no, it doesn't wrt the variable the variation is being taken on.

@hugary1995 is this the correct syntax for using the patch recovery object? I found an out of range error with gdb and I'm not sure where its coming from. I have a user object:

```
  [q_x_sample_gradient_x_patch]
    type = NodalPatchRecoveryMaterialProperty
    patch_polynomial_order = FIRST
    property = 'q_x_sample_gradient_x'
    execute_on = 'NONLINEAR'
    block = sample_material
  []
```

to set up the problem. There is an interface, and so the variable q_x_sample (and the gradient (calculated with a material file)) which this object consumes, is undefined on the other side of the interface. Does the patch recovery object handle that case or do I need to make the mesh disconnected at that interface? I am using the patch recovery to apply interface boundary conditions so the mesh being connected doesn't matter. But I assumed it would work fine since I specified the "block" i.e it should not be attempting to access nodes beyond that block. (edit: can confirm this is not the issue. same error with disconnected mesh).

Then I define an AuxVariable:

```
  [q_x_sample_gradient_x_node]
  []
```

And then apply the patch


```
  [q_x_sample_gradient_x_node]
    type = NodalPatchRecoveryAux
    variable = q_x_sample_gradient_x_node
    nodal_patch_recovery_uo = q_x_sample_gradient_x_patch
    execute_on = 'NONLINEAR'
    block = sample_material
  []
```

This is the backtrace gotten with gdb:

```
(gdb) bt
#0  0x00007fffe8f94820 in PMPI_Abort () from /home/richmond98/mambaforge3/envs/moose/lib/libmpi.so.12
#1  0x00007ffff03a409b in libMesh::libmesh_terminate_handler() () from /home/richmond98/mambaforge3/envs/moose/libmesh/lib/libmesh_dbg.so.0
#2  0x00007fffe93b940c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#3  0x00007fffe93b945e in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:58
#4  0x00007fffe93b9650 in __cxxabiv1::__cxa_throw (obj=0x5555579e5380, tinfo=0x7fffe94d6548 <typeinfo for std::out_of_range>,
    dest=0x7fffe93c622a <std::out_of_range::~out_of_range()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:98
#5  0x00007fffe93b365b in std::__throw_out_of_range (__s=0x7ffff78e2ffe "map::at") at ../../../../../libstdc++-v3/src/c++11/functexcept.cc:86
#6  0x00007ffff7271534 in std::map<unsigned long, Eigen::Matrix<double, -1, -1, 0, -1, -1>, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, Eigen::Matrix<double, -1, -1, 0, -1, -1> > > >::at (this=0x55555657d970, __k=@0x7fffffff9880: 0)
    at /home/richmond98/mambaforge3/envs/moose/x86_64-conda-linux-gnu/include/c++/10.4.0/bits/stl_map.h:550
#7  0x00007ffff722b8ab in NodalPatchRecoveryBase::nodalPatchRecovery (this=0x55555657cb70, x=..., elem_ids=...)
    at /home/richmond98/projects/moose/modules/tensor_mechanics/src/userobjects/NodalPatchRecoveryBase.C:65
#8  0x00007ffff6f6ac39 in NodalPatchRecoveryAux::computeValue (this=0x5555566fe910)
    at /home/richmond98/projects/moose/modules/tensor_mechanics/src/auxkernels/NodalPatchRecoveryAux.C:57
#9  0x00007ffff35980fe in AuxKernelTempl<double>::compute (this=0x5555566fe910)
    at /home/richmond98/projects/moose/framework/src/auxkernels/AuxKernel.C:288
#10 0x00007ffff4693cfa in ComputeNodalAuxVarsThread<AuxKernelTempl<double> >::onNode (this=0x7fffffff9eb0, node_it=...)
    at /home/richmond98/projects/moose/framework/src/loops/ComputeNodalAuxVarsThread.C:98
#11 0x00007ffff32e4f03 in ThreadedNodeLoop<libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*>, __gnu_cxx::__normal_iterator<libMesh::Node const* const*, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > > >::operator() (
    this=0x7fffffff9eb0, range=...) at /home/richmond98/projects/moose/framework/build/header_symlinks/ThreadedNodeLoop.h:109
#12 0x00007ffff32c4fdb in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*>, ComputeNodalAuxVarsThread<AuxKernelTempl<double> > > (range=..., body=...)
    at /home/richmond98/mambaforge3/envs/moose/libmesh/include/libmesh/threads_pthread.h:380
#13 0x00007ffff3255b84 in AuxiliarySystem::computeNodalVarsHelper<AuxKernelTempl<double> > (this=0x5555561f7bb0, warehouse=...)
    at /home/richmond98/projects/moose/framework/src/systems/AuxiliarySystem.C:890
#14 0x00007ffff3250a30 in AuxiliarySystem::computeNodalVars (this=0x5555561f7bb0, type=...)
    at /home/richmond98/projects/moose/framework/src/systems/AuxiliarySystem.C:692
#15 0x00007ffff324dfa8 in AuxiliarySystem::compute (this=0x5555561f7bb0, type=...)
    at /home/richmond98/projects/moose/framework/src/systems/AuxiliarySystem.C:433
#16 0x00007ffff4371c4e in FEProblemBase::computeJacobianTags (this=0x555556152c70, tags=...)
    at /home/richmond98/projects/moose/framework/src/problems/FEProblemBase.C:6516
#17 0x00007ffff4370b9e in FEProblemBase::computeJacobianInternal (this=0x555556152c70, soln=..., jacobian=..., tags=...)
```

**Reply by richmondodufisan** [reply]
This is the block of code where it is failing i.e `/home/richmond98/projects/moose/modules/tensor_mechanics/src/userobjects/NodalPatchRecoveryBase.C:65`


```
  // Before we go, check if we have enough sample points for solving the least square fitting
  if (_q_point.size() * elem_ids.size() < _q)
    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
               "polynomial order or using a higher-order quadrature scheme.");

  // Assemble the least squares problem over the patch
  RealEigenMatrix A = RealEigenMatrix::Zero(_q, _q);
  RealEigenVector b = RealEigenVector::Zero(_q);
  for (auto elem_id : elem_ids)
  {
    A += _Ae.at(elem_id);
    b += _be.at(elem_id);
  }
```

And the variables have the following values:

```
Thread 1 "purple-dbg" hit Breakpoint 1, NodalPatchRecoveryBase::nodalPatchRecovery (this=0x5555563cb280, x=..., elem_ids=...) at /home/richmond98/projects/moose/modules/tensor_mechanics/src/userobjects/NodalPatchRecoveryBase.C:65
65          A += _Ae.at(elem_id);
(gdb) print _q
$1 = 4
(gdb) print elem_ids.size()
Currently Executing
    Computing Automatic Scaling
      Computing Jacobian$2 = 10
(gdb) print _q_point.size()
.$3 = 5
(gdb)
```
